### DISCUSSION TITLE
Using Dexie Cloud for storing Web Push subscriptions?

### DISCUSSION THREAD

**Discussion by naton** [dexie/Dexie.js#1648] [General]
Is anyone using Dexie Cloud to store the subscriptions of Web Push Notifications? 

I'm trying to wrap my head around all the client stuff concerning VAPID keys, subscriptions etc but when it comes to storing the subscriptions on the server people seem to recommend Firebase or FeathersJS – can't we use Dexie in a similar way for this, or am I missing something? Anyone doing this already?

**Comment by naton** [comment]
Just found out about [NeDB](https://github.com/seald/nedb), which [this article](https://web.dev/sending-messages-with-web-push-libraries/) uses for storing web push subscriptions. Is it possible to use Dexie here instead? /ping @dfahlander

**Comment by dfahlander** [comment]
The newly release REST interface can be used from a node app to store key/value items that are kept out from the end users' sync data. It's not exactly covered in the docs how to make the data private for the app, so I'll explain it here (and may copy this text into to the official docs soon).

## Obtain Authorization for Dexie Cloud REST API

Given that you have a client ID and secret (found in dexie-cloud.key file that is generated when you create or connect to a DB using `npx dexie-cloud create` or `npx dexie-cloud connect`), in order to use the API, first request a token for a your node-based 'service account', calling https://yourDB.dexie.cloud/token (https://dexie.org/cloud/docs/rest-api#token) with the parameters:

```js
{
  grant_type: "client_credentials",
  scopes: ["ACCESS_DB"],
  client_id: <your client ID>,
  client_secret: <your client secret>,
  claims: {
    sub: "my-app-service-account@system.local", // Invent arbritary username for the service account
  }
}
```

You'll get back a JSON object with the property "accessToken". Put the content of this in all future requests towards the REST api with header `Authorization: Bearer ${accessToken}`.

## Create a Table (optionally, or use an existing)

To add a table that will only be used by the node application, you'll have to sync the dexie schema and add it command line. Make sure to install the latest version of `dexie-cloud` npm package (current 1.1.0-beta.3), and then:

First, export the current schema that dexie cloud is aware of:
```
npx dexie-cloud export --schema dexie-schema.json
```
Your JSON file will contain something like:

```json
{
  "schema": {
    "members": "@id",
    "roles": "[realmId+name]",
    "realms": "@realmId"
  },
  "sealed": false
}
```

Edit that file and add your new table along it's primary key, such as `"webPushSubscriptions": "id"`.

```json
{
  "schema": {
    "members": "@id",
    "roles": "[realmId+name]",
    "realms": "@realmId",
    "webPushSubscriptions": "id"
  },
  "sealed": false
}
```

Then use dexie-cloud CLI to import the updated schema:

```
npx dexie-cloud import dexie-schema.json
```

## Put an object

[Do a POST request](https://dexie.org/cloud/docs/rest-api#post-and-delete) to `https://yourDB.dexie.cloud/my/webPushSubscriptions`

```
POST /my/webPushSubscriptions HTTP/1.1
Authorization: Bearer <token from /token endpoint>
Content-Type: application/json

[{
  id: <primaryKey>,
  otherProperty: otherValue,
  ...
}]
```
*When POST:in to the `/my/...` endpoint, the realmId will be automatically set to the userId of the caller (which is your service acount) and therfore not accessible by any other user.*

## Get an object back from `https://yourDB.dexie.cloud/my/webPushSubscriptions/<primaryKey>`

```
GET /my/webPushSubscriptions/<primaryKey> HTTP/1.1
Authorization: Bearer <token from /token endpoint>

```

...

I hope this could answer the question on how to use dexie-cloud database from a server-side application for the purpose of storing web push subscriptions.

**Reply by naton** [reply]
Wow, great news and response! Will dive into this more tonight :)

**Reply by naton** [reply]
One thing: Is that beta of `dexie-cloud` installed via some version of `dexie-cloud-addon` or do I have to install it in addition..?

**Reply by dfahlander** [reply]
dexie-cloud is its own npm package (CLI only), apart from dexie-cloud-addon. You actually don't have to install it, since npx will auto-install it temporarily anyway.... Just in case you've installed it earlier, it would be important to upgrade it.

**Comment by dfahlander** [comment]
I've also been sketching on how we could have some kind of push-notification feature within dexie-cloud in the future, or at least web hooks that allows a server application to send notifications based on certain mutations in the database. The road map right now is to release the stable version of the service first before introducing that though.
